### PR TITLE
feat: stepfunction scan queue with semaphore

### PR DESCRIPTION
**Pull request by mohamed-cds**
The API will no longer initiate scans, instead it'll log the request and launch a new scan request via step function. The step function will use dynamodb as a lock table to ensure the established concurrency limit is respected. The step function will keep polling until the result is received and then will release the lock so another scan can begin.

This is required since AssemblyLine only allows 10 active file scan's to be in-process at a time.

Sequence

- User uploads a file by POST to /assemblyline
- The /assemblyline handler will generate a new scan table entry (rds), upload the file to s3 and then associates the two 
- The API will submit new scan requests to the step function scan queue by providing a `scan_id`
- The Step function will then begin execution and will attempt to obtain a lock against the semaphore DynamoDB table
- If no locks are available, the function will sleep and will try again until a lock is obtained
- Once the lock is obtained, the step function will invoke the api lambda and provide a execution_id & scan_id
- The api lambda will launch a new assemblyline scan and provide the execution_id and scan_id as meta-data and will exit
- The step function will now keep polling the api lambda of which any responses from assemblyline will be stored in a dynamodb table containing execution_ids. This is done by evaluating the metadata and storing the execution_id
- Once the step function gets a "True" response from the api lambda indicating that an entry in dynamodb was found containing the current execution_id, the loop is exited and the lock released

image

### CODE REVIEW

File: api/assemblyline/assemblyline.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        f"https://{environ.get('MLWR_HOST')}",
+        timeout=60000,  # Timeout after 1 minute
+        apikey=(environ.get("MLWR_USER"), environ.get("MLWR_KEY")),
+    )
+
+
+def add_to_scan_queue(payload):
+    client = get_session().client("stepfunctions")
+    response = client.list_state_machines()
+    state_machine = "assemblyline-file-scan-queue"
```
Comment (patheard) [inline api/assemblyline/assemblyline.py:23]: minor nit: extract out into file level constant like `dynamodb.DYNAMODB_TABLE `

File: api/assemblyline/assemblyline.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+def poll_for_results():
+
+    try:
+        al_client = get_assemblyline_client()
+        session = next(get_db_session())
+        message = None
+        while True:
+            message = al_client.ingest.get_message("cds_snc_queue")
+            # Results of completed scans are retrieved one at a time until "None" is returned
```
Comment (patheard) [inline api/assemblyline/assemblyline.py:90]: Awesome - thanks for the comment 👍🏻

**Review by patheard — CHANGES_REQUESTED** [review]
I like that I'm now learning step functions by proxy.

Had a few questions/comments, but otherwise looks good!

File: terragrunt/env/scan_queue/state-machines/scan-queue-lock-cleanup.json [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        "TableName": "${table_semaphore}",
+        "ExpressionAttributeNames":  {
+          "#lockowerid.$": "$.detail.executionArn"
+        },
+        "Key": {
+          "LockName": {
+            "S": "${lock_name}"
+          }
+        },
+        "ProjectionExpression": "#lockowerid"
```
[the remark this answers is not in the document]
Comment (mohamed-cds) [reply terragrunt/env/scan_queue/state-machines/scan-queue-lock-cleanup.json]: 100% a typo

File: terragrunt/env/scan_queue/state-machines/scan-queue-lock-cleanup.json [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      "Choices": [
+        {
+          "And": [
+            {
+              "Variable": "$.lockinfo.currentlockitem.ItemString",
+              "IsPresent": true
+            },
+            {
+              "Variable": "$.lockinfo.currentlockitem.ItemString",
+              "StringMatches": "*Z*"
```
[the remark this answers is not in the document]
Comment (mohamed-cds) [reply terragrunt/env/scan_queue/state-machines/scan-queue-lock-cleanup.json:47]: It'll always be a timestamp with the Z so we're safe. If requirements change then it'll be a whole new table. So leaving this as-is.

File: terragrunt/aws/scan_queue/iam.tf [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      "states:ListActivities",
+      "states:CreateStateMachine",
+      "states:CreateActivity",
+      "states:DescribeExecution",
+      "states:StartExecution",
+      "states:StopExecution"
+    ]
+
+    resources = [
+      "arn:aws:states:${var.region}:${var.account_id}:*"
```
[the remark this answers is not in the document]
Comment (mohamed-cds) [reply terragrunt/aws/scan_queue/iam.tf:58]: scoped it down to specific stepfunction so we should be ok

File: terragrunt/aws/api/iam.tf [hunk 6]
```diff
@@ -54,6 +54,53 @@ data "aws_iam_policy_document" "api_policies" {
       "${module.file-queue.s3_bucket_arn}/*",
     ]
   }
+
+  statement {
+
+    effect = "Allow"
+
+    actions = [
+      "states:ListStateMachines",
```
[the remark this answers is not in the document]
Comment (mohamed-cds) [reply terragrunt/aws/api/iam.tf:63]: scoped down to specific stepfunctions

File: api/assemblyline/assemblyline.py (hunk 1 above)
Comment (mohamed-cds) [reply api/assemblyline/assemblyline.py]: moved to environment variable and TF level variable so that i can have it declared in one place.

**Review by patheard — APPROVED** [review]

**Pull request merged**